lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -2759,58 +2791,102 @@ LogicalResult FIRRTLLowering::visitStmt(StopOp op) {
-  auto clock = getLoweredValue(op.clock());
-  auto enable = getLoweredValue(op.enable());
-  auto predicate = getLoweredValue(op.predicate());
-  if (!clock || !enable || !predicate)
+  Value clock, enable, predicate;
+  StringAttr label;
+  StringRef fileName, dir;
+  BOpTy svOp;
+  if (!lowerVerifOps<AOpTy>(op, clock, enable, predicate, label, fileName, dir,
+                            annoClass))
     return failure();
 
   addToAlwaysBlock(clock, [&]() {
     addIfProceduralBlock(enable, [&]() {
       // Create BOpTy inside the always/if.
-      StringAttr label;
-      if (op.nameAttr())
-        label = op.nameAttr();
-      else
-        label = builder.getStringAttr("");
-      auto svOp = builder.create<BOpTy>(predicate, label);
-      auto annoSet = AnnotationSet(circuitState.circuitOp);
-      StringRef fileName, dir;
-      if (auto a = annoSet.getAnnotation(annoClass)) {
-        fileName = a.getAs<StringAttr>("filename").getValue();
-        dir = a.getAs<StringAttr>("directory").getValue();
-      }
-      if (!fileName.empty() || !dir.empty())
-        svOp->setAttr("output_file",
-                      hw::OutputFileAttr::get(builder.getStringAttr(dir),
-                                              builder.getStringAttr(fileName),
-                                              builder.getBoolAttr(true),
-                                              builder.getBoolAttr(true),
-                                              svOp.getContext()));
+      svOp = builder.create<BOpTy>(predicate, label);
     });
   });
 
+  if (!fileName.empty() || !dir.empty())
+    svOp->setAttr("output_file",
+                  hw::OutputFileAttr::get(builder.getStringAttr(dir),
+                                          builder.getStringAttr(fileName),
+                                          builder.getBoolAttr(true),
+                                          builder.getBoolAttr(true),
+                                          svOp.getContext()));
+  return success();
+}
+
+template <typename AOpTy, typename BOpTy>
+LogicalResult
+FIRRTLLowering::lowerVerificationConcurrentStatement(AOpTy op,
+                                                     StringRef annoClass) {
+  Value clock, enable, predicate;
+  StringAttr label;
+  StringRef fileName, dir;
+  BOpTy svOp;
+  if (!lowerVerifOps<AOpTy>(op, clock, enable, predicate, label, fileName, dir,
+                            annoClass))
+    return failure();
+  predicate = builder.createOrFold<comb::AndOp>(enable, predicate);
+  sv::EventControl event;
+  if (op.eventControl() == "AtPosEdge")
+    event = circt::sv::EventControl::AtPosEdge;
+  else if (op.eventControl() == "AtEdge")
+    event = circt::sv::EventControl::AtEdge;
+  else if (op.eventControl() == "AtNegEdge")
+    event = circt::sv::EventControl::AtNegEdge;
+  else {
+    op.emitError(
+        "Event Control can only be one of :'AtPosEdge/AtEdge/AtNegEdge'");
+    return failure();
+  }
```

Comment:
I'm not big on using a string here, but if we have to, this begs for a `StringSwitch` that assigns the `event` directly.

Also, if this was an enumeration of some kind, then we wouldn't need to do verification of the attribute during lowering. 🤔

Reply:
Right, please move to using an enum for this.  There are a bunch of examples in ODS, e.g. `RUWAttr` in FIRRTL dialect

Reply:
Thanks, added a FIRRTL enum attribute to fix this.